src/extractor/edge_based_graph_factory.cpp

```diff
@@ -449,8 +450,47 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
             }(turn_classification.second);
             bearing_class_by_node_based_node[node_at_center_of_intersection] = bearing_class_id;
 
-            for (const auto &turn : intersection)
+            const EdgeWeight traffic_light_penalty = [&]() {
+                if (m_traffic_lights.find(node_at_center_of_intersection) != m_traffic_lights.end())
+                {
+                    return profile_properties.traffic_signal_penalty;
+                }
+                return 0;
+            }();
+
+            const auto turn_to_penalty =
+                [this, &scripting_environment, traffic_light_penalty](const auto &turn) -> int32_t {
+                    // don't add turn penalty if it is not an actual turn. This heuristic is
+                    // necessary since OSRM cannot handle looping roads/parallel roads
+                    if (turn.instruction.type == guidance::TurnType::NoTurn)
+                        return 0;
+
+                    int32_t turn_penalty =
+                        scripting_environment.GetTurnPenalty(180. - turn.angle);
+
+                    if (turn.instruction.direction_modifier == guidance::DirectionModifier::UTurn)
+                    {
+                        turn_penalty += profile_properties.u_turn_penalty;
+                    }
+
+                    turn_penalty += traffic_light_penalty;
+
+                    return turn_penalty;
+                };
+
+            std::vector<int32_t> turn_penalties(intersection.size());
+            std::transform(intersection.begin(),
+                           intersection.end(),
+                           turn_penalties.begin(),
+                           turn_to_penalty);
+            const int64_t total_turn_penalties = std::accumulate(turn_penalties.begin(), turn_penalties.end(), 0);
+            const int32_t average_turn_penalty = total_turn_penalties / intersection.size();
+
+            for (const auto turn_index : util::irange<std::size_t>(0, intersection.size()))
             {
+                const auto &turn = intersection[turn_index];
+                const auto turn_penalty = turn_penalties[turn_index];
+
                 // only keep valid turns
                 if (!turn.entry_allowed)
```

Comment:
what are the implications of having `invalid` turns in the set we compute `turn_penalties` on? Can invalid instructions (I am thinking of the currently miss-used `DirectionModifier::UTurn` for invalid turns) modify the normalisation for the worse?